ndsl/testing/comparison.py

```diff
@@ -1,68 +1,265 @@
-    ignore_near_zero_errors: Union[dict, bool],
-    near_zero: float,
-):
-    success = np.logical_or(
-        np.logical_and(np.isnan(computed_data), np.isnan(ref_data)),
-        compare_arr(computed_data, ref_data) < eps,
-    )
-    if isinstance(ignore_near_zero_errors, dict):
-        if ignore_near_zero_errors.keys():
-            near_zero = ignore_near_zero_errors["near_zero"]
+
+    def __init__(
+        self,
+        reference_values: np.ndarray,
+        computed_values: np.ndarray,
+        eps: float,
+        ignore_near_zero_errors: Union[dict, bool],
+        near_zero: float,
+    ):
+        super().__init__(reference_values, computed_values)
+        self.eps = eps
+        self.success = self._compute_errors(
+            ignore_near_zero_errors,
+            near_zero,
+        )
+        self.check = np.all(self.success)
+        self._calculated_metric = np.empty_like(self.references)
+
+    def _compute_errors(
+        self,
+        ignore_near_zero_errors,
+        near_zero,
+    ) -> npt.NDArray[np.bool_]:
+        if self.references.dtype in (np.float64, np.int64, np.float32, np.int32):
+            denom = np.abs(self.references) + np.abs(self.computed)
+            self._calculated_metric = np.asarray(
+                2.0 * np.abs(self.computed - self.references) / denom
+            )
+            self._calculated_metric[denom == 0] = 0.0
+        elif self.references.dtype in (np.bool_, bool):
+            self._calculated_metric = np.logical_xor(self.computed, self.references)
+        else:
+            raise TypeError(
+                f"recieved data with unexpected dtype {self.references.dtype}"
+            )
+        success = np.logical_or(
+            np.logical_and(np.isnan(self.computed), np.isnan(self.references)),
+            self._calculated_metric < self.eps,
+        )
+        if isinstance(ignore_near_zero_errors, dict):
+            if ignore_near_zero_errors.keys():
+                near_zero = ignore_near_zero_errors["near_zero"]
+                success = np.logical_or(
+                    success,
+                    np.logical_and(
+                        np.abs(self.computed) < near_zero,
+                        np.abs(self.references) < near_zero,
+                    ),
+                )
+        elif ignore_near_zero_errors:
             success = np.logical_or(
                 success,
                 np.logical_and(
-                    np.abs(computed_data) < near_zero,
-                    np.abs(ref_data) < near_zero,
+                    np.abs(self.computed) < near_zero,
+                    np.abs(self.references) < near_zero,
                 ),
             )
-    elif ignore_near_zero_errors:
-        success = np.logical_or(
-            success,
-            np.logical_and(
-                np.abs(computed_data) < near_zero, np.abs(ref_data) < near_zero
-            ),
-        )
-    return success
+        return success
 
+    def __str__(self) -> str:
```

Comment:
As per [docs]( https://docs.python.org/3/reference/datamodel.html#object.__repr__) :
If a class defines __repr__ but not __str__(), then __repr__() is also used when an “informal” string representation of instances of that class is required.

I think is not necessary to do this, but does not hurt

Reply:
My blurry mind seem to remember an instance where this mattered. But it might be under py 2...